base/sort.jl

```diff
@@ -1625,14 +1631,26 @@ function sortperm!(ix::AbstractArray{T}, A::AbstractArray;
                    scratch::Union{Vector{T}, Nothing}=nothing,
                    dims...) where T <: Integer #to optionally specify dims argument
     (typeof(A) <: AbstractVector) == (:dims in keys(dims)) && throw(ArgumentError("Dims argument incorrect for type $(typeof(A))"))
-    axes(ix) == axes(A) || throw(ArgumentError("index array must have the same size/axes as the source array, $(axes(ix)) != $(axes(A))"))
+    axes(ix) == axes(A) || @noinline throw_sortperm_axis_mismatch_error(ix, A)
 
     if !initialized
-        ix .= LinearIndices(A)
+        very_unsafe_copyto!(ix, LinearIndices(A))
+    end
+
+    if rev === true
+        sort!(ix; alg, order=Perm(ord(lt, by, true, order), vec(A)), scratch, dims...)
+    else
+        sort!(ix; alg, order=Perm(ord(lt, by, nothing, order), vec(A)), scratch, dims...)
     end
-    sort!(ix; alg, order = Perm(ord(lt, by, rev, order), vec(A)), scratch, dims...)
 end
 
+# TODO stop using these three hacks
+# but check performance, especially unexpected allocations, when removing
+Base.@assume_effects :nothrow very_unsafe_copyto!(a, b) = a .= b
```

Comment:
this is scary

Reply:
Yeah... @aviatesk, is this sort of thing acceptable or is there another way?

Reply:
My local test shows that perhaps `@noinline` is enough?

Reply:
This is why we do code review. Thanks!

I think I introduced these when they were necessary and then added branching on `rev` which is more impactful and fixes the root of the issue so now none of these hacks are necessary at all, and CI checks for unexpected allocations automatically :)